kore/src/Kore/Step/SMT/Translate.hs

```diff
@@ -228,28 +314,43 @@ translatePredicateWith translateTerm predicate =
                 -- will fail to translate.
                 SMT.not <$> translateBool notChild
             ApplySymbolF app ->
-                translateApplication (Just SMT.tBool) pat app
-            DefinedF (Defined child) -> translateBool child
+                translateApplication (Just SMT.tBool) pat' app
+            DefinedF (Defined child) ->
+                withDefinednessAssumption
+                $ translateBool child
             _ -> empty
 
-    translateApplication :: Maybe SExpr -> p -> Application Symbol p -> Translator m variable SExpr
+    translateApplication
+        :: Maybe SExpr
+        -> TermLike variable
+        -> Application Symbol (TermLike variable)
+        -> Translator variable monad SExpr
     translateApplication
         maybeSort
         original
         Application
             { applicationSymbolOrAlias
             , applicationChildren
             }
-      | isFunctionalPattern original =
-        translateInterpretedApplication
-        <|> translateUninterpreted'
-      | otherwise =
-        translateInterpretedApplication
+      = do
+        TranslatorEnv { assumeDefined } <- ask
+        translateApplicationWorker assumeDefined
       where
+        translateApplicationWorker isDefined
+          | (isDefined && isFunctionPattern original)
+          || isFunctionalPattern original =
+              translateInterpretedApplication
+              <|> translateUninterpreted'
+          | otherwise =
+              -- TODO: this is not right, function-like
+              -- symbols should be sent to the solver only if
+              -- we know they are defined. Arbitrary symbols
+              -- should never be sent to the solver.
+              translateInterpretedApplication
```

Comment:
```suggestion
      =
        -- TODO: This would send interpreted symbols to the solver
        -- even if they may not be defined. We should only send symbols
        -- we know to be defined.
        translateInterpretedApplication
        -- TODO: Move call to guardLocalFunctionalPattern inside translateUnintepreted'.
        <|> (guardLocalFunctionalPattern >> translateUninterpreted')
        
        translateApplicationWorker assumeDefined
      where
        guardLocalFunctionalPattern
          | isFunctionalPattern original = return ()
          | otherwise = do
            TranslatorEnv { assumeDefined } <- ask
            Monad.guard (assumeDefined && isFunctionPattern original)
```